### PR TITLE
Electronic scale for Z axis probe

### PR DESCRIPTION
**Pull request by necdnk**
This sensor very sensitive. You can see it: http://www.youtube.com/watch?v=MtTQW5f7Cek
But sensitive can be changed. 
This solution is very precision (near 0,02mm on my 3d printer).
Description of implementation in Russian:
http://geektimes.ru/post/249138/

Algorithm of the calibration:
1. Set the sensor's sensitivity about 300 grams
2. Execute the first (hard) detection.
3. Set the sensor's sensitivity of about 5-50 grams.
4. Execute a second (accurate) detection.

### CODE REVIEW

**Comment by oysteinkrog** [comment]
How well does it handle plastic on the nozzle?
Does it work with bed surfaces that are not conductive? (e.g. covered in abs juice).

**Comment by necdnk** [comment]
You can use any surface (glass, metal, Kapton, ABS juice)
In this video, i use mirror:
http://www.youtube.com/watch?v=b3wcx6VSzj4
If you are using ABS juice, you can put in the slicer z-offset height value of this layer.
Starting the calibration must be done on the heated extruder. In this case, you get a very precise adjustment of the height.
Ask questions, I'm happy to answer them.
Sorry for my English ;)

File: Marlin/Configuration_adv.h [hunk 1]
```diff
@@ -192,7 +192,7 @@
 #define X_HOME_BUMP_MM 5
 #define Y_HOME_BUMP_MM 5
 #define Z_HOME_BUMP_MM 2
-#define HOMING_BUMP_DIVISOR {2, 2, 4}  // Re-Bump Speed Divisor (Divides the Homing Feedrate)
+#define HOMING_BUMP_DIVISOR {2, 2, 40}  // Re-Bump Speed Divisor (Divides the Homing Feedrate)
```
Comment (tonnico) [inline Marlin/Configuration_adv.h]: don't change this in the standard configs

File: Marlin/Marlin_main.cpp [hunk 2]
```diff
@@ -54,6 +54,7 @@
 #include "language.h"
 #include "pins_arduino.h"
 #include "math.h"
+#include "HX711.h"
```
Comment (tonnico) [inline Marlin/Marlin_main.cpp]: #ifdef ELECTRONIC_SCALE_PROBE?

**Comment by tonnico** [comment]
You must update your fork before merge. The endstop logic changed some weeks ago. There are no booleans anymore for the endstops.

**Comment by thinkyhead** [comment]
@necdnk You'll step all over your own toes making changes directly to your Development branch. You should always make a new branch for PRs. For example, `git checkout Development -b my_feature_branch`. But before you do that you'll need to get the freshest Development branch. For example, `git fetch upstream ; git checkout Development ; git merge upstream/Development`.

**Comment by oysteinkrog** [comment]
I think this is a pretty nice way to do distance measurement, I especially like that the measurement is done with the actual nozzle itself (without applying too much force etc).
I think I will order the parts and try is myself.
@necdnk Is there a reason you used the 5kg load cell instead of the 3kg?
I would think 3kg (or maybe the 1kg model too) should be enough?
http://www.aliexpress.com/item/Kitchen-Digital-Electronic-Scale-Pressure-Sensor-Load-Cell-YZC-131-3Kg-3-Hole/32320250538.html

**Comment by necdnk** [comment]
@oysteinkrog yes! you can try to use 3kg load cell. I think this load cells (3kg and 5kg) is equals. 
You can try using 1kg sensor, but it is advisable not to exceed the pressure in 2kg.
I pulled it out of, from destroyed kitchen scales. But later I bought several sensors from China, and this is also working successfully.
I want to try the tensoresistors used in these sensors like this: http://ru.aliexpress.com/item/Free-shipping-5pcs-lot-BF350-3AA-BF350-Precision-resistive-strain-gauge-strain-gauge-for-the-pressure/1935464167.html
They are very small, but need to learn how to use them correctly.

**Comment by necdnk** [comment]
@thinkyhead I will need to make a new PR? From the new branch? And this PR to be closed?
I apologize for such questions. This is my first PR, before I did not work with them.

**Comment by Wackerbarth** [comment]
@necdnk -- No need for a new PR. Just rebase the code to the current head of the MarlinFirmware Development branch, make any revisions needed, and push it to your github account. Automatically, this PR will be updated by the github infrastructure.

For your future reference, when starting a new feature or patch, start a new branch within your repository, one per issue. That way you can easily work on multiple issues at the same time. When the issue is closed, you should delete the branch so that you don't clutter up your branch namespace with out-of-date, now irrelevant, labels.

**Comment by thinkyhead** [comment]
My preferred method to fix up your type of situation would be:

```
git fetch upstream
git checkout Development
git merge upstream/Development ;# merges are usually easier
git checkout upstream/Development -b my_clean_feature
git diff my_clean_feature Development | git apply
```

Then you can cherry-pick the feature branch to create separate commits, or just make one large commit.

At this point your Development branch will still be out of sync with upstream, so you will want to reset it to match the upstream Development HEAD.

**Comment by Wackerbarth** [comment]
"or just make one large commit." -- Actually, breaking your commit into smaller pieces has advantages.
(1) For someone reviewing the code, smaller pieces are easier to understand.
(2) Conflict resolution is also easier to manage.

**Comment by thinkyhead** [comment]
> breaking your commit into smaller pieces has advantages

Truly. I particularly like how cherry-picking allows you to group changes that should go together, which otherwise the diff algorithm might not be able to figure out.

File: Marlin/Configuration.h [hunk 3]
```diff
@@ -544,6 +544,14 @@ const bool Z_PROBE_ENDSTOP_INVERTING = false; // set to true to invert the logic
 
 #endif // ENABLE_AUTO_BED_LEVELING
 
+//Enable weight sensor (load cell) for Z axis.
+//#define ELECTRONIC_SCALE_PROBE
+#ifdef ELECTRONIC_SCALE_PROBE
+	#define SCALE_GAIN 64 // 128 or 64 only for A channel 32 only for B channel
+	#define SCALE_SENSIVITY_HARD 40
+	#define SCALE_SENSIVITY_ACCURATE 5
+	#define SCALE_DIVISOR 1000.f
```
Comment (thinkyhead) [inline Marlin/Configuration.h]: Two spaces, no tabs.

http://www.marlinfirmware.org/index.php/Coding_Standards

File: Marlin/HX711.h [hunk 4]
```diff
@@ -0,0 +1,88 @@
+#ifndef __HX711_h
+#define __HX711_h
+
+#if ARDUINO >= 100
+#include "Arduino.h"
```
Comment (thinkyhead) [inline Marlin/HX711.h:4]: Indent items between `#if ... #endif`.

http://www.marlinfirmware.org/index.php/Coding_Standards

**Comment by boelle** [comment]
Please rebase, the code cant be merged as is

**Comment by necdnk** [comment]
This code tested on my printer and ready for merge.
Please check it.

**Comment by thinkyhead** [comment]
The changes to these files seem strange, in part because the diffs are flagging lines that seem to be the same:

```
ArduinoAddons/Arduino_1.6.x/hardware/marlin/avr/libraries/U8glib/utility/u8g_circle.c
ArduinoAddons/Arduino_1.6.x/hardware/marlin/avr/libraries/U8glib/utility/u8g_dev_ssd1309_128x64.c
```

**Comment by necdnk** [comment]
@thinkyhead I think it's a mistake comparing git :(
I merge changes from upstream/Development into my repository.

**Comment by oysteinkrog** [comment]
You need to rebase this, it's a mess right now.

**Comment by thinkyhead** [comment]
@necdnk Are you able to redo this? You'll need to create a separate branch (not named "Development") and rebuild your feature there.

So first make a copy of your branch with its commits:

```
git checkout Development -b my_feature_branch
```

Then you might do:

```
git remote add upstream https://github.com/MarlinFirmware/Marlin.git ;# if needed
git fetch upstream
git checkout Development
git reset --hard upstream/Development
```

And finally:

```
git checkout my_feature_branch
git rebase Development
```

To fix up your fork's `Development` branch as stored on Github you will need to change your fork's default branch to something other than `Development` using the [Settings link for your fork](https://github.com/necdnk/Marlin/settings), then delete the `Development` branch, and then finally push your local `Development` branch to `origin/Development`. Something along the lines of:

```
git checkout Development
git branch --unset-upstream
git push origin/Development
```

**Pull request closed without merging**